pkg/servicescm/servicescm.go

```diff
@@ -125,6 +134,150 @@ func Generate(name, namespace string) (*core.ConfigMap, error) {
 	return servicesConfigMap, nil
 }
 
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(dataFromCM map[string]string) (*data, error) {
+	if len(dataFromCM) != 2 {
+		return nil, errors.New("services ConfigMap should have exactly 2 keys")
+	}
+
+	services := &[]Service{}
+	files := &[]FileInfo{}
+
+	value, ok := dataFromCM[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", servicesKey)
+	}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, err
+	}
+
+	value, ok = dataFromCM[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected key %s does not exist", filesKey)
+	}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, err
+	}
+
+	return newData(services, files)
+}
+
+// validate ensures the given object represents a valid services ConfigMap, made up of only required services and files.
+// The validation also ensures bootstrap services are defined to always start before controller services.
+func (cmData *data) validate() error {
+	if err := validateRequiredContent(cmData.Services, cmData.Files); err != nil {
+		return err
+	}
+	if err := validateDependencies(cmData.Services); err != nil {
+		return err
+	}
+	return validatePriorities(cmData.Services)
+}
+
+func validateRequiredContent(services []Service, files []FileInfo) error {
+	// Validate services
+	if len(services) != len(*requiredServices) {
+		return errors.New("Unexpected number of services")
+	}
+	for _, requiredSvc := range *requiredServices {
+		// isPresent checks if the required service exists within the given services slice
+		isPresent := func() error {
+			for _, svc := range services {
+				if reflect.DeepEqual(requiredSvc, svc) {
+					return nil
+				}
+			}
+			return errors.Errorf("Required service %s is not present with expected configuration", requiredSvc.Name)
+		}
```

Comment:
You are defining this function per required service. Won't it be more efficient to make this a helper?

Reply:
Yup I noticed this after opening for review so didn't fix it right away :). I'll pull these out into helper on the struct types.